cmake/modules/dts.cmake

```diff
@@ -1,5 +1,12 @@
 # SPDX-License-Identifier: Apache-2.0
 
+include_guard(GLOBAL)
+
+# Dependencies of this module.
+include(extensions)
+include(boards)
+include(generic_toolchain)
+
```

Comment:
No, build configuration itself is not a direct requirement, but a valuable support module.

What this means is that one could use the `dts.cmake` with another set of configuration files specified before loading this module.

`build_configuration` just provides a support module making it easier to call `dts.cmake` for a Zephyr build.
A use-case where one could want to use `dts.cmake` without the `build_configuration` could be #40555.

As the `dts.cmake` is an existing file, I have not added extra documentation.
This will follow in a dedicated CMake doc PR.
Doc PRs can generate a lot of comments, which is why I only added description for purely new modules added.

Reply:
Is there any way to tell which modules are requirements and which are support modules?

Theoretically, any direct requirement could be replaced with user-supplied code.

Can the module system supply alternative module implementations, just like you can have multiple header files with the same name, and modify just the include path?

Reply:
This was answered elsewhere.

Reply:
> Is there any way to tell which modules are requirements and which are support modules?
> 

1. Seen from the module itself ? like inside `boards.cmake` is this a support module, or required module
2. Or from the using module ?  like `kconfig.cmake`, do I require `boards.cmake` or is it just a support module.

For 1. I wouldn't know how a module itself can determine if someone else consider it a required module.

For 2., all required modules should be listed at the top of the CMake module, and support modules should be described in the general text of the module. (Currently working on followup PR that will add description text to all old modules that are just relocated with this PR)

Roughly speaking, one can say that an undefined `BOARD_ROOT` will not break the system, there will just not be any boards available, whereas a missing CMake function from `extensions.cmake` will fail like:
```
CMake Error at cmake/modules/zephyr_module.cmake:33 (set_ifndef):                                                                                                                                                                            
  Unknown CMake command "set_ifndef".    
```
or a missing Python from `python.cmake` will make code like this to fail:
https://github.com/zephyrproject-rtos/zephyr/blob/1c57242423cfd3280ed742f44ed33bfd66cbc72d/cmake/modules/kconfig.cmake#L155
https://github.com/zephyrproject-rtos/zephyr/blob/1c57242423cfd3280ed742f44ed33bfd66cbc72d/cmake/modules/kconfig.cmake#L262

Of course their might be cases where an undefined CMake variable itself will also break things, and that's why required CMake variables should be described in the module description, but for now I find the split between functions and program for required, and pure CMake variables sensible.
I expect that there is a higher wish to control CMake variables when re-using Zephyr CMake modules, than to overwrite functions or programs required for the Zephyr build system.

> Theoretically, any direct requirement could be replaced with user-supplied code.

It could, but this is not even supported today, so I do not intend to support this use-case straight away.
With this PR we prepend the Zephyr module path to `CMAKE_MODULE_PATH` to ensure Zephyr CMake modules are looked up first, as to ensure thinks has been tested together.
https://github.com/zephyrproject-rtos/zephyr/blob/1c57242423cfd3280ed742f44ed33bfd66cbc72d/share/zephyr-package/cmake/ZephyrConfig.cmake#L14
This is similar to what CMake itself is doing:
in future we can consider if we should relax this 
> 
> Can the module system supply alternative module implementations, just like you can have multiple header files with the same name, and modify just the include path?

Reply:
Since this is an internal module we can ensure that build_configuration is included before dts.cmake when appropriate.

Reply:
except that for re-usability, a project reusing `dts.cmake` might want a different scheme for build configuration files.